api/txspool.go

```diff
@@ -38,6 +40,79 @@ func (a *API) postPoolTx(c *gin.Context) {
 	c.JSON(http.StatusOK, writeTx.TxID.String())
 }
 
+func (a *API) postAtomicPool(c *gin.Context) {
+	// Parse body
+	var receivedTxs []receivedPoolTx
+	if err := c.ShouldBindJSON(&receivedTxs); err != nil {
+		retBadReq(err, c)
+		return
+	}
+	nTxs := len(receivedTxs)
+	if nTxs <= 1 {
+		retBadReq(errors.New(ErrSingleTxInAtomicEndpoint), c)
+		return
+	}
+	// Transform from received to insert format and validate (individually)
+	writeTxs := make([]l2db.PoolL2TxWrite, nTxs) // used for DB insert
+	txIDStrings := make([]string, nTxs)          // used for successful response
+	clientIP := c.ClientIP()
+	for i, tx := range receivedTxs {
+		writeTx := tx.toPoolL2TxWrite()
+		if err := a.verifyPoolL2TxWrite(writeTx); err != nil {
+			retBadReq(err, c)
+			return
+		}
+		writeTx.ClientIP = clientIP
+		writeTxs[i] = *writeTx
+		txIDStrings[i] = writeTx.TxID.String()
+	}
+	// Validate that all txs in the payload represent an atomic group
+	if !isAtomicGroup(writeTxs) {
+		retBadReq(errors.New(ErrTxsNotAtomic), c)
+		return
+	}
+	// Insert to DB
+	if err := a.l2.AddAtomicTxsAPI(writeTxs); err != nil {
+		retSQLErr(err, c)
+		return
+	}
+	// Return IDs of the added txs in the pool
+	c.JSON(http.StatusOK, txIDStrings)
+}
+
+// isAtomicGroup returns true if all the txs are needed to be forged
+// (all txs will be forged in the same batch or non of them will be forged)
+func isAtomicGroup(txs []l2db.PoolL2TxWrite) bool {
+	// Create a graph from the given txs to represent requests between transactions
+	g := graph.New(len(txs))
+	idToPos := make(map[common.TxID]int, len(txs))
+	// Map tx ID to integers that will represent the nodes of the graph
+	for i, tx := range txs {
+		idToPos[tx.TxID] = i
+	}
+	// Create vertices that connect nodes of the graph (txs) using RqTxID
+	for i, tx := range txs {
+		if tx.RqTxID == nil {
```

Comment:
whats happen if only one tx doesn't have a `RqTxID` parameter

Reply:
The idea of this endpoint is that all submitted txs are tomic within the same group (all the txs sended in the request have to be forged together). Therefore if only one tx doesn't request any other, this tx could be forged alone, making the group not 100% atomic.

Of course this is an arbitrary criteria that I made up, and we can change it. Let me share my reasoning:

- Someone could sneak txs pretending to be part of the atomic group using other people's fees to subvention their own txs (of course we could add logic to cover this case, but this way we avoid this extra logic)
- Prevent users from doing not intended things: atomic txs are not that easy to reason about, if we've an endpoint dedicated to atomic txs my guess is that people will send txs to this endpoint with the idea of this txs will all get forged or non of them will get forged. And this is exactly what I've implemented